File: c_src/cerck.c
/* Copyright (c) 2010, Michael Santos <michael.santos@gmail.com>
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 
 * Redistributions of source code must retain the above copyright
 * notice, this list of conditions and the following disclaimer.
 * 
 * Redistributions in binary form must reproduce the above copyright
 * notice, this list of conditions and the following disclaimer in the
 * documentation and/or other materials provided with the distribution.
 * 
 * Neither the name of the author nor the names of its contributors
 * may be used to endorse or promote products derived from this software
 * without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
#include <crack.h>
#include <string.h>
#include "erl_nif.h"


static ERL_NIF_TERM atom_ok;
static ERL_NIF_TERM atom_error;
static ERL_NIF_TERM atom_enomem;
static ERL_NIF_TERM string_dictpath;

/* cracklib is not thread safe */
typedef struct {
    ErlNifMutex *lock;
} PRIV;

static ERL_NIF_TERM error_tuple(ErlNifEnv *env, char *err);


    static int
load(ErlNifEnv *env, void **priv_data, ERL_NIF_TERM load_info)
{
    PRIV *priv = NULL;

    atom_ok = enif_make_atom(env, "ok");
    atom_error = enif_make_atom(env, "error");
    atom_enomem = enif_make_atom(env, "enomem");

    string_dictpath = enif_make_string(env, GetDefaultCracklibDict(), ERL_NIF_LATIN1);

    priv = (PRIV *)enif_alloc(sizeof(PRIV));
    if (priv == NULL)
        return (-1);

    priv->lock = enif_mutex_create("cerck_lock");
    if (priv->lock == NULL)
        return (-1);

    *priv_data = priv;

    return (0);
}

    void
unload(ErlNifEnv *env, void *priv_data)
{
    PRIV *priv = NULL;

    priv = (PRIV *)enif_priv_data(env);
    enif_mutex_destroy(priv->lock);
    enif_free(priv);
}  

/* 1: password, 2: dictpath */
    static ERL_NIF_TERM
nif_check(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])
{
    PRIV *priv = NULL;

    ErlNifBinary passwd;
    ErlNifBinary path;

    char *err = NULL;


    priv = (PRIV *)enif_priv_data(env);

    if (!enif_inspect_iolist_as_binary(env, argv[0], &passwd))
        return enif_make_badarg(env);

    if (!enif_inspect_iolist_as_binary(env, argv[1], &path))
        return enif_make_badarg(env);

    /* NULL terminate strings */
    if (!enif_realloc_binary(&passwd, passwd.size+1))
        return atom_enomem;
    if (!enif_realloc_binary(&path, path.size+1))
        return atom_enomem;

    /* passwd.size is now equal to old passwd.size+1 */
    passwd.data[passwd.size-1] = '\0';
    path.data[path.size-1] = '\0';

    enif_mutex_lock(priv->lock);
    err = (char *)FascistCheck((char *)passwd.data, (char *)path.data);
    enif_mutex_unlock(priv->lock);

    (void)memset(passwd.data, '\0', passwd.size);

    enif_release_binary(&passwd);
    enif_release_binary(&path);

    return ( (err == NULL) ? atom_ok : error_tuple(env, err));
}

    static ERL_NIF_TERM
nif_dictpath(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])
{
    return string_dictpath;
}

    static ERL_NIF_TERM
error_tuple(ErlNifEnv *env, char *err)
{
    return enif_make_tuple(env, 2,
            atom_error,
            enif_make_string(env, err, ERL_NIF_LATIN1));
}  


static ErlNifFunc nif_funcs[] = {
    {"check", 2, nif_check},
    {"dictpath", 0, nif_dictpath}
};

ERL_NIF_INIT(cerck, nif_funcs, load, NULL, NULL, unload)

File: src/cerck.erl
%%% @copyright 2010-2021, Michael Santos <michael.santos@gmail.com>

%% All rights reserved.
%%
%% Redistribution and use in source and binary forms, with or without
%% modification, are permitted provided that the following conditions
%% are met:
%%
%% Redistributions of source code must retain the above copyright
%% notice, this list of conditions and the following disclaimer.
%%
%% Redistributions in binary form must reproduce the above copyright
%% notice, this list of conditions and the following disclaimer in the
%% documentation and/or other materials provided with the distribution.
%%
%% Neither the name of the author nor the names of its contributors
%% may be used to endorse or promote products derived from this software
%% without specific prior written permission.
%%
%% THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
%% "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
%% LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
%% FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
%% COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
%% INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
%% BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
%% LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
%% CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
%% LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
%% ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
%% POSSIBILITY OF SUCH DAMAGE.
-module(cerck).
-include_lib("cerck/include/cerck.hrl").

-export([
    dictpath/0,
    check/1, check/2,
    quality/1,
    has/2, has/3
]).

-on_load(on_load/0).

-type passwd_quality() :: #passwd_quality{}.
-export_type([passwd_quality/0]).

on_load() ->
    erlang:load_nif(niflib(), []).

% @doc Get the cracklib dictionary path
%
% Returns the base path and filename prefix to the cracklib database.
%
-spec dictpath() -> string().
dictpath() ->
    erlang:nif_error(not_implemented).

% @doc Test for weak or predictable passwords, specifying the dictionary path
-spec check(iodata(), iodata()) -> ok | {error, string()}.
check(_, _) ->
    erlang:nif_error(not_implemented).

% @doc Test for weak or predictable passwords
%
% ```
% 1> cerck:check(<<"foobar">>).
% {error,"it is based on a dictionary word"}
% 2> cerck:check(<<"f00b4r">>).
% ok
% '''
-spec check(iodata()) -> ok | {error, string()}.
check(Passwd) ->
    check(Passwd, dictpath()).

% @doc Analyze a password
-spec quality(string() | binary()) -> passwd_quality().
quality(Passwd) when is_binary(Passwd) ->
    quality(binary_to_list(Passwd));
quality(Passwd) when is_list(Passwd) ->
    quality1(Passwd, #passwd_quality{}).

quality1([], Stats) ->
    Stats;
quality1([H | T], #passwd_quality{lower = N} = Stats) when H >= $a, H =< $z ->
    quality1(T, Stats#passwd_quality{lower = N + 1});
quality1([H | T], #passwd_quality{upper = N} = Stats) when H >= $A, H =< $Z ->
    quality1(T, Stats#passwd_quality{upper = N + 1});
quality1([H | T], #passwd_quality{number = N} = Stats) when H >= $0, H =< $9 ->
    quality1(T, Stats#passwd_quality{number = N + 1});
quality1([_ | T], #passwd_quality{other = N} = Stats) ->
    quality1(T, Stats#passwd_quality{other = N + 1}).

% @doc Test password for character types
-spec has(
    'lower' | 'number' | 'other' | 'upper',
    passwd_quality()
) -> boolean().
has(lower, Stats) ->
    has(lower, Stats, 1);
has(upper, Stats) ->
    has(upper, Stats, 1);
has(number, Stats) ->
    has(number, Stats, 1);
has(other, Stats) ->
    has(other, Stats, 1).

% @doc Test password for character types specifying a minimum number
-spec has(
    'lower' | 'number' | 'other' | 'upper',
    passwd_quality(),
    non_neg_integer()
) -> boolean().
has(lower, #passwd_quality{lower = N}, Min) when N >= Min -> true;
has(upper, #passwd_quality{upper = N}, Min) when N >= Min -> true;
has(number, #passwd_quality{number = N}, Min) when N >= Min -> true;
has(other, #passwd_quality{other = N}, Min) when N >= Min -> true;
has(Type, #passwd_quality{}, _) when
    Type == lower; Type == upper; Type == number; Type == other
->
    false.

privdir(File) ->
    filename:join([
        filename:dirname(code:which(?MODULE)),
        "..",
        "priv",
        File
    ]).

niflib() ->
    privdir(?MODULE).
